app/Http/Controllers/CategoryController.php

```diff
@@ -0,0 +1,148 @@
+
+        return view('admin.categories.index', compact('categories'));
+    }
+
+    /**
+     * Show the form for creating a new resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function create()
+    {
+        return view('admin.categories.create');
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(CategoryCreate $request)
+    {
+        $params = $request->all();
+        $params['slug'] = Str::slug($params['name']);
+
+        $category = $this->categoryRepo->create($params);
+
+        if (blank($category)) {
+            Session::flash('error', 'Thêm mới thất bại');
+        } else {
+            Session::flash('success', 'Thêm mới thành công');
+        }
+
+        return back();
+    }
+
+    /**
+     * Display the specified resource.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function show($id)
+    {
+        //
+    }
+
+    /**
+     * Show the form for editing the specified resource.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function edit($id)
+    {
+        $category = $this->categoryRepo->getById($id);
+
+        if (blank($category)) {
+            Session::flash('error', 'Không tìm thấy dữ liệu !');
+            return back();
+        }
+
+        return view('admin.categories.edit', compact('category'));
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function update(CategoryUpdate $request, $id)
+    {
+        $params = $request->all();
+        unset($params['_token'], $params['_method']);
+        $params['slug'] = Str::slug($params['name']);
+
+        $updation = $this->categoryRepo->update($id, $params);
+
+        if (!$updation) {
+            Session::flash('error', 'Cập nhật thất bại !');
+        } else {
+            Session::flash('success', 'Cập nhật thành công !');
+        }
+
+        return back();
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy($id)
+    {
+        $destroy = $this->categoryRepo->clear($id);
+
+        switch ($destroy) {
+            case -1:
+                Session::flash('error', 'Không tìm thấy dữ liệu !');
+                break;
+            case 0:
+                Session::flash('error', 'Xóa thất bại !');
+                break;
+            case 1:
+                Session::flash('success', 'Xóa thành công !');
+                break;
+            default:
+                Session::flash('error', 'Xóa thất bại !');
+                break;
+        }
```

Comment:
Đoạn này xử lý rườm rà nhỉ sao không xử lý bớt trong ham clear, hàm clear cg chỉ dùng cho trường hợp này thôi mà

Reply:
Em không xử lý trong hàm clear vì em muốn những hàm trong Repositories chỉ được phép thực hiện truy vấn dữ liệu và trả kết quả cho câu truy vấn đó, còn việc xử lý kết quả đó như thế nào thì phải xử lý bến Controller, em nghĩ như thế sẽ tường minh - dễ quản lý hơn.

Reply:
nếu hàm clear như vậy thif không khác gì làm phức tạp hơn, không nhất thiết phải tách ra như em đang lamf, có thể viết hàm find và xử lý luôn trong controller

Reply:
Từ hôm qua em cố tìm hiểu xem làm sao để bớt rườm rà như anh nói. Nhưng có vẻ trước giờ em toàn làm theo kiểu đó giờ quen, nên đối với em nó gọn và em không tìm được cách. Em có thể được biết nếu là anh thì anh sẽ làm sao để cho nó tốt hơn. Nếu xử lý trong hàm clear thôi thì xử lý ra sao mà trong controller thì ra sao

Reply:
Nếu xử lý trong hàm clear thì xử lý 3 trường hợp rồi trả về message kèm status luôn để khi gọi trong controller thì không cần phải if else nữa
Nếu xử lý luôn trong controler thì đơn giản là tách hàm find ra xử lý và handler luôn việc không tìm thấy dữ liệu, hàm này có thể tái sử dụng trong các hàm show, update, ... Còn việc xử lý không xóa đc hay không update đc sẽ đc xử lý ở chính hàm destroy, update ...